USAGOV-354-Update-Node-Modules

cwacht opened the pull request:
Switch from uswds-gulp to uswds-compile and switch node from alpine to bullseye

<!--- Provide a general summary of your changes in the title above -->
## Jira Task
<!--- Provide a link to the Jira ticket -->
https://cm-jira.usa.gov/browse/USAGOV-354

## Description
<!--- Describe your changes in detail -->

## Checklist for the Developer
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask for help! -->
- [ ] A link to the JIRA ticket has been added above.
- [ ] The JIRA ticket identifies the desired result of this change.
- [ ] The JIRA ticket contains clear acceptance criteria.
- [ ] The JIRA ticket contains clear testing steps.
- [ ] The JIRA ticket contains a description of "Done".
- [ ] Any preparation/installation/update steps required for this code to work properly (drush commands, scripts, configuration updates) are documented in the ticket.

## Checklist for the Peer Reviewers
- [ ] The branch name of this PR matches the project standards.
- [ ] QA steps are followed and any changes to process are updated in the ticket.
- [ ] Code Standards are followed, there are no bad practices in use.
- [ ] Config changes (if any) include only necessary changes.
- [ ] No errors in Drupal or Client Browser.
- [ ] Code has been tested locally (if possible).
- [ ] Following AC and Testing Steps verify changes work as expected.
- [ ] There are no known side-effects outside of expected behavior.

cwacht committed:
USAGOV-354-Update-Node-Modules: switch from uswds-gulp to uswds-compile and switch node from alpine to bullseye

cwacht committed:
USAGOV-354-Update-Node-Modules: Modern javascript linting

akf committed:
Merge branch 'dev' into USAGOV-354-Update-Node-Modules

akf commented:
I intend to review this, soon! We'll need to either dial down the javascript linting or fix some things in our js -- or both. Probably the same applies for css.

IsabelLaurenceau commented:
![Screen Shot 2023-01-30 at 10 54 34 AM](https://user-images.githubusercontent.com/108532577/215526816-bc682312-631d-4ca9-93d5-6f38a428b3aa.png)
![Screen Shot 2023-01-30 at 10 54 27 AM](https://user-images.githubusercontent.com/108532577/215526818-4c823ca7-8dc7-4446-81cb-a5421e909f7c.png)
![Screen Shot 2023-01-30 at 10 54 20 AM](https://user-images.githubusercontent.com/108532577/215526819-04da716e-7a54-47f8-b992-3be7ad828069.png)

Looks like something was off with the styling when I made a new build from this branch this morning.

I didn't get any error messages in the terminal; the logs looked similar to other builds I do

akf committed:
Merge branch 'dev' into USAGOV-354-Update-Node-Modules

akf commented:
I'm seeing pages similar to what @IsabelLaurenceau does. The USWDS compilation generates a lot fewer messages than the old version, which probably doesn't mean anything.

cwacht committed:
USAGOV-354-Update-Node-Modules: remove package-lock.json

cwacht committed:
USAGOV-354-Update-Node-Modules: Merge branch 'dev' into USAGOV-354-Update-Node-Modules # Conflicts: # web/themes/custom/usagov/package-lock.json

cwacht committed:
USAGOV-354-Update-Node-Modules: run theme's build command instead of init

cwacht committed:
Merge branch 'dev' into USAGOV-354-Update-Node-Modules

cwacht committed:
USAGOV-354-Update-Node-Modules: ignore JS files that we don't edit

cwacht committed:
USAGOV-354-Update-Node-Modules: adjust which files get linted

cwacht commented:
With the 2 commits that I made today, I think I've gotten my local linting output to match what I see on Circle CI.
Question, do I need to fix all the linting errors before we can approve this pull request? If so, should I fix them in this pull request?

akf commented:
I think they should all be fixed (or the linting rules tweaked further -- it looks like we can downgrade specific rules from "error" pretty easily in eslint, I didn't look at stylelint). Basically, the linter needs to pass; otherwise we'll just learn to ignore it. 
That doesn't mean you need to do all the work! We can split this up. If you want to do that, it's probably best if people make branches off of this one and you merge them in. I'll sign up for at least some of the js work. 

I see this at the end of the report: 

"125 errors and 0 warnings potentially fixable with the `--fix` option." 

Maybe running "npx eslint --fix scripts/*" would be a good first step?

cwacht committed:
USAGOV-354-Update-Node-Modules: automatic lint fixes

akf commented on .docker/Dockerfile-cms:
```diff
@@ -75,20 +75,20 @@ RUN sed -E -i "s/:x:${BUID}:/:x:1919:/g" /etc/passwd \
-
-COPY bin/src/phantomjs.sh /local/phantomjs.sh
-RUN . /local/phantomjs.sh
+# RUN apk upgrade --update \
```
Let's delete the commented-out code.

akf commented:
Just FYI: I need to look into this further ... 

This still has the problem where a fresh build (in my case, using the pr-review script, which is a bit suspect) winds up missing the styles -- more of them than last time! "git diff" shows me this: 

	modified:   web/themes/custom/usagov/package-lock.json
	modified:   web/themes/custom/usagov/sass/_uswds-theme-custom-styles.scss
	modified:   web/themes/custom/usagov/sass/_uswds-theme.scss

It looks like both of the scss files have been reverted to what a fresh USWDS build would have, probably. 

I'm looking at my latest notes on the Jira ticket and not quite following something I wrote about how we're invoking "npm init," and I'm not tracking whether / where we're invoking that now.

cwacht committed:
USAGOV-354-Update-Node-Modules: manual lint fixes

cwacht commented:
My fix to prevent those files from being overridden was to change `bin/npm run init` to `bin/npm run build` in the /bin/init file.

My best guess is that if you were running the pr-review script while on any other branch, then your /bin/init file probably still ran `bin/npm run init` which would cause those files to get overridden.

akf committed:
Merge branch 'dev' into USAGOV-354-Update-Node-Modules

akf commented:
I'm probably confused about something -- bin/init on the dev branch doesn't have a call to "bin/npm run init" (anymore?) that I can see. 
I'm going to try this from a fresh clone of the repo, though.

cwacht commented:
It looks like you're right about that. I dont know what else would have overwritten those files. Make sure that those changes are not present when you start a fresh build.

akf commented:
On a totally fresh installation (cleared out all my docker images, even), I'm getting a different error and my build doesn't complete -- IIRC it failed to build the node image as well as a couple of other things. The error message mentions making sure I have "sass-embedded-linux-arm64" installed, which makes me think this might be an issue that hits M1 Macs only. Here's the error, plus a couple of lines above. I'm going to try to figure out what I might need to add here, but I thought I'd capture it: 

```
#0 11.74 > usagov_sass@1.0.0 build
#0 11.74 > npx gulp compile
#0 11.74 
#0 12.40 Error: Embedded Dart Sass couldn't find the embedded compiler executable. Please make sure the optional dependency sass-embedded-linux-arm64 is installed in node_modules.
#0 12.40     at /var/www/web/themes/custom/usagov/node_modules/sass-embedded/dist/lib/src/compiler-path.js:28:11
#0 12.40     at Object.<anonymous> (/var/www/web/themes/custom/usagov/node_modules/sass-embedded/dist/lib/src/compiler-path.js:32:3)
#0 12.40     at Module._compile (node:internal/modules/cjs/loader:1218:14)
#0 12.40     at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
#0 12.40     at Module.load (node:internal/modules/cjs/loader:1081:32)
#0 12.40     at Module._load (node:internal/modules/cjs/loader:922:12)
#0 12.40     at Module.require (node:internal/modules/cjs/loader:1105:19)
#0 12.40     at require (node:internal/modules/cjs/helpers:103:18)
#0 12.40     at Object.<anonymous> (/var/www/web/themes/custom/usagov/node_modules/sass-embedded/dist/lib/src/async-compiler.js:10:25)
#0 12.40     at Module._compile (node:internal/modules/cjs/loader:1218:14)
```

akf commented:
Confirmed -- this is an M1 Mac vs. Intel issue. I was able to get up and running by removing the package-lock.json file and running bin/init again; this time the node build pulled in arm64-specific dependencies. No other files were modified! 

I think we want the package-lock file, though. So I'm noodling on how best to handle this.

akf commented on .docker/Dockerfile-cms:
```diff
@@ -65,7 +65,7 @@ RUN COMPOSER_MEMORY_LIMIT=-1 composer install --ignore-platform-reqs --no-intera
 # USER www-data
 ###############################################################
-FROM ${NODE_BASE}:18-alpine3.16 AS theme-builder
+FROM ${NODE_BASE}:19.3-bullseye AS theme-builder
```
Change 19.3 to 19.9 (newer version with fewer vulnerabilities)

akf committed:
USAGOV-354-multiarch-package-lock: Update package-lock.json with all optional dependencies.

File was created by deleting *both* package-lock.json and node_modules
and doing an npm install (via bin/init).

akf committed:
Merge branch 'dev' into USAGOV-354-Update-Node-Modules

akf commented:
Apparently there is an npm bug (or perhaps there's disagreement over whether it's a bug) that results in different package-lock.json contents depending on whether you already have node_modules installed when you install or update packages. I've created a pull request to merge into _this branch_ and tagged you as a reviewer -- please see if that works on your machine, and if so, merge it in.

akf reviewed (changes requested):
Question: did you add "use strict" to the javascript functions because eslint seemed to be asking for it? I think we should make sure to test all the modified javascript just in case. 

Beyond that, I think I'm done reviewing. If my change to the package-lock.json file works for you, I would approve this with that PR merged in. Would like to aim to merge this shortly after a deployment to Staging (so, on a Friday) to allow for regression testing on dev.

cwacht committed:
Merge pull request #890 from usagov/USAGOV-354-multiarch-package-lock

USAGOV-354: Update package-lock.json with all optional dependencies.

cwacht committed:
Update version number and remove commented code

cwacht committed:
Merge branch 'dev' into USAGOV-354-Update-Node-Modules

cwacht committed:
USAGOV-354-Update-Node-Modules: add this branch for dev deployment

akf committed:
USAGOV-354-use-updateUSWDS: Use uswds.updateUswds for theme rebuilds

akf committed:
USAGOV-354-use-updateUSWDS: Keep the "npm run build" command and make it point to uswdsUpdate

akf committed:
USAGOV-354-use-updateUSWDS: removed accidentally added init.bak

akf reviewed (changes requested):
I've made another pull request for a change that fixed the missing files problem on my local dev site: https://github.com/usagov/usagov-2021/pull/893. (Doing the PR partly to make it easy to track the changes, but also because I think if I commit to this branch I can't approve it anymore!)

I don't know why this would have worked for some builds, and my best guess is that we weren't really starting from scratch when we had "success" before. 

I recommend @IsabelLaurenceau  as a third reviewer. Another test deployment to the dev space will definitely be a good review step as well.

cwacht committed:
USAGOV-354-Update-Node-Modules: During init, allow our theme files to be overwritten, then copy them back before build

cwacht committed:
USAGOV-354-Update-Node-Modules: ignore the files that get created upon init

cwacht commented:
OK I just pushed some changes that do a different fix, but I like your way better, so I will revert my changes and merge yours in. I'm honestly confused about why your solution is not the default starting point for USWDS compile...

cwacht committed:
Revert "USAGOV-354-Update-Node-Modules: ignore the files that get created upon init"

This reverts commit 92f2c6abb95992d560a8f993e7ee44c8623912ae.

cwacht committed:
Revert "USAGOV-354-Update-Node-Modules: During init, allow our theme files to be overwritten, then copy them back before build"

This reverts commit 845320228358022e8dc873865a9918ef220a8a18.

cwacht committed:
Merge pull request #893 from usagov/USAGOV-354-use-updateUSWDS

USAGOV-354: use uswds.updateUswds

cwacht committed:
Revert "USAGOV-354-Update-Node-Modules: add this branch for dev deployment"

This reverts commit ae11b0308a2665b099f02600f45638f482d04237.

akf committed:
Merge branch 'dev' into USAGOV-354-Update-Node-Modules

akf reviewed (approved):
Now working on your machine, my machine, and dev!
